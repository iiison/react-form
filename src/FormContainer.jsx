import React, { Component } from 'react'
import PropTypes            from 'prop-types' // eslint-disable-line import/no-extraneous-dependencies

/**
 * Main Component, a HOC, will store all fields' state.
 */
export default class FormContainer extends Component {
  constructor(props) {
    super(props)

    const {
      isDisabled = false,
      shouldValidateForm = false
    } = this.props

    const defaultState = {
      isDisabled         : false,
      shouldValidateForm : false
    }

    this.state = {
      ...defaultState,
      shouldValidateForm,
      isDisabled
    }
  }

  getChildContext() {
    return {
      setFieldValue : ({ event, field, value, isMultipleValues }) => {
        const fieldName = field.id
        const setState = (name, fieldValue) => {
          this.setState((prevState) => ({
            ...prevState,
            fields : {
              ...prevState.fields,
              [fieldName] : {
                ...prevState.fields[name],
                value : fieldValue
              }
            }
          }))
        }

        if (!value) {
          setState(fieldName, event.currentTarget.value)
        } else if (isMultipleValues) {
          const stateCopy = { ...this.state }

          for (const item in value) {
            stateCopy.fields[item].value = value[item]
          }

          this.setState((prevState) => ({
            ...prevState,
            ...stateCopy
          }))
        } else {
          setState(fieldName, value)
        }

        return this.state
      },

      addField : (data) => {
        this.setState((prevState) => ({
          ...prevState,
          fields : {
            ...prevState.fields,
            [data.id] : data 
          }
        }))
      },

      validateForm : (fieldName) => {
        const { shouldValidateForm, fields } = this.state

        if (shouldValidateForm) {
          if (fieldName) {
            FormContainer.validateField(fields[fieldName])

            return
          }

          for (const field in fields) {
            if (field.validate) {
              this.validateField(fields[field])
            }
          }
        }

        return
      },
      formData : this.state
    }
  }

  /**
   * React Lifecycle Method: Renders the data
   *
   * @return {DOM} Main container DOM.
   */
  render() {
    return (
      <form action=''>
        {this.props.children}
      </form>
    )
  }

  static validateField = (fieldData) => {
    const rules = fieldData.validate.split('|')

    console.log(fieldData)
  }

  static defaultProps = {
    children           : <div />,
    isDisabled         : false,
    shouldValidateForm : false
  }

  static propTypes = {
    children           : PropTypes.node.isRequired,
    isDisabled         : PropTypes.bool,
    shouldValidateForm : PropTypes.bool
  }

  static childContextTypes = {
    addField      : PropTypes.func,
    setFieldValue : PropTypes.func,
    validateForm  : PropTypes.func,
    formData      : PropTypes.object
  }
}

