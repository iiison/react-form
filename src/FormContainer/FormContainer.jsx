import React, { Component } from 'react'

import { Form } from '$COMPONENTS'
import validations from '$UTILS/validationRules'

/**
 * Test all validation rules ofa field
 *
 * @param {Object} fieldData all field realated data
 *
 * @return {String}         validation error ina field
 */
export function testFieldRules(fieldData) {
  const { validate, value, displayName } = fieldData

  const validationRules = validate.split('|')

  for (const validation in validationRules) {
    const { rule, formatter } = validations[validationRules[validation]]()

    const result = rule.test(value)

    if (result === false) {
      return formatter(displayName)
    }
  }
}

/**
 * Generic Form Handler, Handles form state,
 * Pass Form elemnt's JSON throught the FormContainer and
 * it will redner the form.
 */
export default class FormContainer extends Component {
  /**
   * Constructor for the container
   *
   * @param {Object} props form data
   */
  constructor(props) {
    super(props)
    const { formData : { fields } } = props

    this.state = {
      isTested : false,
    }

    this.state.fields = fields.reduce((prev, curr) => {
      const temp = {
        value          : '',
        error          : '',
        shouldBeTested : false,
        ...curr
      }

      return {
        ...prev,
        [temp.id] : temp
      }
    }, {})
  }

  /**
   * DOM Event : Chaanges `shouldBeTested`, will be executed on
   * keyUp in an input field
   *
   * @param {String} field field name
   */
  makeFieldTestable = (field) => {
    const fieldRef = this.state.fields[field]

    fieldRef.shouldBeTested = true

    this.setState((previousState) => {
      return {
        fields : {
          ...previousState.fields,
          [field] : fieldRef
        }
      }
    })
  }

  /**
   * Handle Key press event on the input,
   * tests field value if field is already has error.
   * @param  {String} field Field name in which event was started
   */
  handleKeyPress = (field) => {
    const fieldData = this.state.fields[field]
    const { error } = fieldData

    this.makeFieldTestable(field)

    if (error) {
      this.validateInput(field)
    }
  }

  /**
   * DOM Event: Handles input value change
   *
   * @param {Object} event DOM event object
   */
  handleChange = (event) => {
    event.preventDefault()

    const elemRef = event.target
    const idName = elemRef.id

    this.setState((previousState) => {
      const { fields } = previousState

      return {
        fields : {
          ...fields,
          [idName] : {
            ...fields[idName],
            value : elemRef.value
          }
        }
      }
    })
  }

  /**
   * Validate single input
   *
   * @param  {String}  field       field name to be tested
   * @param  {Boolean} isFinalTest will be true if whole form is being tested
   * @return {String}              validation error of the field
   */
  validateInput = (field, isFinalTest) => {
    const fieldData = this.state.fields[field]
    const { validate } = fieldData

    if ((fieldData.shouldBeTested === true || isFinalTest) && validate) {
      const error = testFieldRules(fieldData) || ''

      this.setState((previousState) => {
        const { fields } = previousState
        const updatedField = {
          ...fields[field],
          error
        }

        return {
          fields : {
            ...fields,
            [field] : updatedField
          }
        }
      })

      return error
    }
  }

  /**
   * Handles Form Submission, callback actual form submit fxn
   * after the form validation. validation is the part of plugin.
   *
   * @param {Object} event DOM Event object
   */
  handleFormSubmit = (event) => {
    event.preventDefault()

    const { handleSubmit } = this.props
    const { fields } = this.state
    const errors = []

    for ( const field in fields ) {
      const error = this.validateInput(field, true)

      error && errors.push(error)
    }

    if (!errors.length) {
      handleSubmit(fields)
    }
  }

  /**
   * React Lifecycle Method: Renders the data
   *
   * @return {DOM} Main container DOM.
   */
  render() {
    const { fields } = this.state
    const {
      error,
      messages,
      isFetching,
      submitButtonText,
      submitButtonClassName } = this.props

    const props = {
      fields,
      messages,
      events : {
        onSubmit          : this.handleFormSubmit,
        onValueChange     : this.handleChange,
        validateInput     : this.validateInput,
        handleKeyPress    : this.handleKeyPress,
        makeFieldTestable : this.makeFieldTestable
      },
      submitButtonData : {
        error,
        isFetching,
        submitButtonText,
        submitButtonClassName,
        onSubmit : this.handleFormSubmit
      }
    }

    return (
      <Form
        {...props}
      />
    )
  }

  /**
  * React Lifecycle Method: Executes right before the
  * component mounts
  */
  componentWillMount() {
  }
}
