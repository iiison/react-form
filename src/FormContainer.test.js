import React, { Component } from 'react'
import renderer      from 'react-test-renderer'
import { mount }     from 'enzyme'
import PropTypes     from 'prop-types' // eslint-disable-line import/no-extraneous-dependencies

import FormContainer from './FormContainer'

/* eslint-disable react/jsx-filename-extension */
class SmallComponent extends Component {
  componentDidMount() {
  }

  render() {
    return (
      <div>
        <input type='text' />
      </div>    
    )
  }
}

SmallComponent.contextTypes = {
  addField : PropTypes.func
}

// Snapshot matching for Form Container
describe('>>> Form Container -- Snapshot Test', () => {
  it('Matches the snapshot when no elements passed to the Form', () => {
    const tree = renderer.create(<FormContainer />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Matches the snapshot when some data is passed to the Form', () => {
    const formContent = (
      <div>
        <input type='text' />
      </div>
    )
    const tree = renderer.create(
      <FormContainer>
        {formContent}
      </FormContainer>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
// *************************************************************

/* eslint-enable */


// Check DOM Behaviour
describe('>>> Form Container -- Shallow Rendering', () => {
  const formContent = (
    <div>
      <input type='text' />
    </div>
  )
  let mountedScreen
  const formScreen = (customState = {}, content) => {
    if (!mountedScreen) {
      mountedScreen = mount(
        <FormContainer {...customState}>
          {content}
        </FormContainer>
      )
    }

    return mountedScreen
  }

  beforeEach(() => {
    mountedScreen = undefined
  })

  it('Renders Empty Form', () => {
    const tree = formScreen({}, [])

    const expected = 0
    const result = tree
      .find('form')
      .children()
      .length

    expect(result).toEqual(expected)
  })

  it('Renders Form Component With valid elements in it', () => {
    const tree = formScreen({}, <SmallComponent />)

    // TODO: Need to test context variables, still figuringout a way to access them here.
  })

  // TODO: State Test when function called
})
// *************************************************************

