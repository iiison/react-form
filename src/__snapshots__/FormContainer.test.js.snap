// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Form Container -- Snapshot Test Matches the snapshot when no elements passed to the Form 1`] = `
<form
  action=""
>
  <div />
</form>
`;

exports[`>>> Form Container -- Snapshot Test Matches the snapshot when some data is passed to the Form 1`] = `
<form
  action=""
>
  <div>
    <input
      type="text"
    />
  </div>
</form>
`;
